File: java/src/test/java/scenario/TransactionInvocation.java
/*
 * Copyright 2019 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package scenario;

import java.security.GeneralSecurityException;
import java.util.Map;
import java.util.concurrent.Callable;

import org.hyperledger.fabric.client.Commit;
import org.hyperledger.fabric.client.CommitStatusException;
import org.hyperledger.fabric.client.Contract;
import org.hyperledger.fabric.client.EndorseException;
import org.hyperledger.fabric.client.Gateway;
import org.hyperledger.fabric.client.GatewayException;
import org.hyperledger.fabric.client.Proposal;
import org.hyperledger.fabric.client.Status;
import org.hyperledger.fabric.client.SubmitException;
import org.hyperledger.fabric.client.SubmittedTransaction;
import org.hyperledger.fabric.client.Transaction;
import org.hyperledger.fabric.client.identity.Signer;

import static org.assertj.core.api.Assertions.assertThat;

public final class TransactionInvocation {
    private final Gateway gateway;
    private final Proposal.Builder proposalBuilder;
    private Callable<byte[]> action;
    private Signer offlineSigner;
    private String response;
    private Throwable error;
    private long blockNumber;

    private TransactionInvocation(final Gateway gateway, final Contract contract, final String transactionName) {
        this.gateway = gateway;
        proposalBuilder = contract.newProposal(transactionName);
    }

    public void setTransient(final Map<String, String> transientData) {
        transientData.forEach(proposalBuilder::putTransient);
    }

    public void setEndorsingOrgs(final String[] orgs) {
        proposalBuilder.setEndorsingOrganizations(orgs);
    }

    public static TransactionInvocation prepareToSubmit(final Gateway gateway, final Contract contract, final String transactionName) {
        TransactionInvocation invocation = new TransactionInvocation(gateway, contract, transactionName);
        invocation.action = invocation::submit;
        return invocation;
    }

    public static TransactionInvocation prepareToEvaluate(final Gateway gateway, final Contract contract, final String transactionName) {
        TransactionInvocation invocation = new TransactionInvocation(gateway, contract, transactionName);
        invocation.action = invocation::evaluate;
        return invocation;
    }

    public void setArguments(final String[] args) {
        proposalBuilder.addArguments(args);
    }

    public void setOfflineSigner(final Signer signer) {
        offlineSigner = signer;
    }

    public void invoke() {
        try {
            byte[] result = action.call();
            setResponse(result);
        } catch (Exception e) {
            setError(e);
        }
    }

    private byte[] submit() throws GeneralSecurityException, EndorseException, SubmitException, CommitStatusException {
        Proposal unsignedProposal = proposalBuilder.build();
        Proposal signedProposal = offlineSign(unsignedProposal);

        Transaction unsignedTransaction = signedProposal.endorse();
        Transaction signedTransaction = offlineSign(unsignedTransaction);

        SubmittedTransaction submitted = signedTransaction.submitAsync();
        Commit commit = offlineSign(submitted);

        Status status = commit.getStatus();
        blockNumber = status.getBlockNumber();

        if (!status.isSuccessful()) {
            throw new RuntimeException("Transaction commit failed with status: " + status.getCode());
        }

        return submitted.getResult();
    }

    private Proposal offlineSign(final Proposal proposal) throws GeneralSecurityException {
        if (null == offlineSigner) {
            return proposal;
        }

        byte[] signature = offlineSigner.sign(proposal.getDigest());
        return gateway.newSignedProposal(proposal.getBytes(), signature);
    }

    private Transaction offlineSign(final Transaction transaction) throws GeneralSecurityException {
        if (null == offlineSigner) {
            return transaction;
        }

        byte[] signature = offlineSigner.sign(transaction.getDigest());
        return gateway.newSignedTransaction(transaction.getBytes(), signature);
    }

    private Commit offlineSign(final Commit commit) throws GeneralSecurityException {
        if (null == offlineSigner) {
            return commit;
        }

        byte[] signature = offlineSigner.sign(commit.getDigest());
        return gateway.newSignedCommit(commit.getBytes(), signature);
    }

    private byte[] evaluate() throws GeneralSecurityException, GatewayException {
        Proposal unsignedProposal = proposalBuilder.build();
        Proposal signedProposal = offlineSign(unsignedProposal);

        return signedProposal.evaluate();
    }

    private void setResponse(final byte[] response) {
        this.response = ScenarioSteps.newString(response);
        error = null;
    }

    private void setError(final Throwable error) {
        this.error = error;
        response = null;
    }

    public String getResponse() {
        assertThat(response)
                .withFailMessage(() -> "No transaction response. Failed with error: " + error)
                .isNotNull();
        return response;
    }

    public Throwable getError() {
        assertThat(error)
                .withFailMessage(() -> "No transaction error. Response was: " + response)
                .isNotNull();
        return error;
    }

    public long getBlockNumber() {
        return blockNumber;
    }
}

File: scenario/node/src/transactioninvocation.ts
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Contract, Gateway, ProposalOptions, Signable, Signer, StatusCode } from '@hyperledger/fabric-gateway';
import { bytesAsString, toError, toString } from './utils';

export class TransactionInvocation {
    readonly options: ProposalOptions = {};

    private readonly name: string;
    private readonly gateway: Gateway;
    private readonly contract: Contract;
    private readonly invoke: () => Promise<Uint8Array>;
    private offlineSigner?: Signer;
    private result?: Uint8Array;
    private error?: Error;
    private blockNumber = BigInt(0);

    constructor(type: string, gateway: Gateway, contract: Contract, name: string) {
        this.gateway = gateway;
        this.contract = contract;
        this.name = name;
        this.invoke = this.getInvoke(type).bind(this);
    }

    async invokeTransaction(): Promise<void> {
        try {
            this.result = await this.invoke();
        } catch (error) {
            this.error = toError(error);
        }
    }

    setOfflineSigner(signer: Signer): void {
        this.offlineSigner = signer;
    }

    getResult(): string {
        if (!this.result) {
            throw new Error(`No transaction result. Error is: ${toString(this.error)}`);
        }

        return bytesAsString(this.result);
    }

    getError(): Error {
        if (!this.error) {
            throw new Error(`No transaction error. Result is: ${bytesAsString(this.result)}`);
        }

        return this.error;
    }

    getBlockNumber(): bigint {
        return this.blockNumber;
    }

    private getInvoke(type: string): () => Promise<Uint8Array> {
        if (type === 'evaluate') {
            return () => this.evaluate();
        }
        if (type === 'submit') {
            return () => this.submit();
        }
        throw new Error(`Unknown invocation type: ${type}`);
    }

    private async evaluate(): Promise<Uint8Array> {
        let proposal = this.contract.newProposal(this.name, this.options);
        proposal = await this.sign(proposal, this.gateway.newSignedProposal.bind(this.gateway));

        return await proposal.evaluate();
    }

    private async submit(): Promise<Uint8Array> {
        const unsignedProposal = this.contract.newProposal(this.name, this.options);
        const signedProposal = await this.sign(unsignedProposal, this.gateway.newSignedProposal.bind(this.gateway));

        const unsignedTransaction = await signedProposal.endorse();
        const signedTransaction = await this.sign(
            unsignedTransaction,
            this.gateway.newSignedTransaction.bind(this.gateway),
        );

        const submitted = await signedTransaction.submit();
        const signedCommit = await this.sign(submitted, this.gateway.newSignedCommit.bind(this.gateway));

        const status = await signedCommit.getStatus();

        this.blockNumber = status.blockNumber;

        if (status.code !== StatusCode.VALID) {
            throw new Error(`Transaction commit failed with status: ${status.code}`);
        }

        return submitted.getResult();
    }

    private async sign<T extends Signable>(
        signable: T,
        newInstance: (bytes: Uint8Array, signature: Uint8Array) => T,
    ): Promise<T> {
        if (!this.offlineSigner) {
            return signable;
        }

        const signature = await this.offlineSigner(signable.getDigest());
        return newInstance(signable.getBytes(), signature);
    }
}
